core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
     }
 
 protected:
+    /***
+     * Aggregate the amount of samples that can be consumed/produced from a range of ports.
+     * @param ports a typelist of input or output ports
+     * @return a tuple representing the amount of available data on the ports
+     *   - minSync: the minimum amount of samples that the block needs for processing on the sync ports
+     *   - maxSync: the maximum amount of that can be consumed on all sync ports
+     *   - maxAvailable: the maximum amount of that are available on all sync ports
+     *   - hasAsync: true if there is at least one async input/output that has available samples/remaining capacity
+     */
+    auto
+    getPortLimits(auto &&ports) {
+        std::size_t minSync               = 0UL;
+        std::size_t maxSync               = std::numeric_limits<std::size_t>::max();
+        std::size_t maxAvailable          = std::numeric_limits<std::size_t>::max();
+        bool        hasAsync              = false;
+        auto        adjust_for_input_port = [this, &minSync, &maxSync, &maxAvailable, &hasAsync]<PortLike Port>(Port &port) {
+            const std::size_t available = [&port]() {
+                if constexpr (gr::traits::port::is_input_v<Port>) {
+                    return port.streamReader().available();
+                } else {
+                    return port.streamWriter().available();
+                }
+            }();
+            if constexpr (std::remove_cvref_t<Port>::kIsSynch) {
+                minSync      = std::max(minSync, port.min_samples);
+                maxSync      = std::min(maxSync, port.max_samples);
+                maxAvailable = std::min(maxAvailable, available);
+            } else {                                 // async port
+                if (available >= port.min_samples) { // ensure that process function is called if at least one async port has data available
+                    hasAsync = true;
+                }
+            }
+        };
+        for_each_port([&adjust_for_input_port](PortLike auto &port) { adjust_for_input_port(port); }, std::forward<decltype(ports)>(ports));
+        return std::tuple{minSync, maxSync, maxAvailable, hasAsync};
```

Comment:
Would it be better if one returns a struct instead of tuple?
```cpp
struct PortLimits {
   std::size_t minSync{};
   std::size_t maxSync{};
   std::size_t maxAvailable{};
   bool hasAsync{};
};
```
And later one can do:
```cpp
const auto in = getPortLimits(inputPorts<PortType::STREAM>(&self()));
const auto out = getPortLimits(outputPorts<PortType::STREAM>(&self()));
```